src/main/java/uk/gov/companieshouse/api/accounts/service/impl/DirectorsReportServiceImpl.java

```diff
@@ -0,0 +1,249 @@
+        }
+
+        if (entity == null) {
+            return new ResponseObject<>(ResponseStatus.NOT_FOUND);
+        }
+        return new ResponseObject<>(ResponseStatus.FOUND, directorsReportTransformer.transform(entity));
+    }
+
+    @Override
+    public ResponseObject<DirectorsReport> delete(String companyAccountsId, HttpServletRequest request)
+            throws DataException {
+        String reportId = generateID(companyAccountsId);
+
+        try {
+            if (directorsReportRepository.existsById(reportId)) {
+
+                directorsReportRepository.deleteById(reportId);
+
+                smallFullService.removeLink(companyAccountsId, SmallFullLinkType.DIRECTORS_REPORT, request);
+                return new ResponseObject<>(ResponseStatus.UPDATED);
+            } else {
+
+                return new ResponseObject<>(ResponseStatus.NOT_FOUND);
+            }
+        } catch (MongoException e) {
+
+            throw new DataException(e);
+        }
+    }
+
+    @Override
+    public void addDirector(String companyAccountsID, String directorID, String link, HttpServletRequest request)
+            throws DataException {
+
+        String reportId = generateID(companyAccountsID);
+        DirectorsReportEntity entity = directorsReportRepository.findById(reportId)
+                .orElseThrow(() -> new DataException(
+                        "Failed to get Directors report entity from which to add director"));
+        if (entity.getData().getDirectorsEntity() == null) {
+
+            entity.getData().setDirectorsEntity(new HashMap<>());
+        }
+        entity.getData().getDirectorsEntity().put(directorID, link);
+
+        try {
+
+            directorsReportRepository.save(entity);
+        } catch (MongoException e) {
+
+            throw new DataException(e);
+        }
+    }
+
+    @Override
+    public void removeDirector(String companyAccountsID, String directorID, HttpServletRequest request)
+            throws DataException {
+        String reportId = generateID(companyAccountsID);
+
+        DirectorsReportEntity entity = directorsReportRepository.findById(reportId)
+                .orElseThrow(() -> new DataException(
+                        "Failed to get directors report entity from which to remove link"));
+
+        entity.getData().getDirectorsEntity().remove(directorID);
+
+        try {
+
+            directorsReportRepository.save(entity);
+        } catch (MongoException e) {
+
+            throw new DataException(e);
+        }
+
+    }
+
+    @Override
+    public void addSecretary(String companyAccountsID, String directorID, String link, HttpServletRequest request) throws DataException {
+        String reportId = generateID(companyAccountsID);
+        DirectorsReportEntity entity = directorsReportRepository.findById(reportId)
+                .orElseThrow(() -> new DataException(
+                        "Failed to get Directors report entity from which to add director"));
```

Comment:
Careful on the wording of these... 
Maybe
'Failed to get directors report entity to which to add director / secretary'
and
'Failed to get directors report entity from which to remove director / secretary'?